src/DeviceInterface/InterfacePrototype.js

```diff
@@ -537,9 +513,87 @@ class InterfacePrototype {
-    supportsFeature (_name) {
-        return false
+    /** @param {FeatureToggleNames} name */
+    supportsFeature (name) {
+        return this.supportedFeatures.includes(name)
+    }
+
+    /**
+     * `preAttachTooltip` happens just before a tooltip is show - features may want to append some data
+     * at this point.
+     *
+     * For example, if password generation is enabled, this will generate
+     * a password and send it to the tooltip as though it were a stored credential.
+     *
+     * @param {TopContextData} topContextData
+     * @param {HTMLInputElement} input
+     * @param {{isSignup: boolean|null}} form
+     */
+    preAttachTooltip (topContextData, input, form) {
+        // A list of checks to determine if we need to generate a password
+        const checks = [
+            topContextData.inputType === 'credentials.password',
+            this.supportsFeature('password.generation'),
+            form.isSignup
+        ]
+
+        // if all checks pass, generate and save a password
+        if (checks.every(Boolean)) {
+            const password = this.passwordGenerator.generate({
+                input: input.getAttribute('passwordrules'),
+                domain: window.location.hostname
+            })
+
+            // append the new credential to the topContextData so that the top autofill can display it
+            topContextData.credentials = [fromPassword(password)]
+        }
+
+        return topContextData
+    }
+
+    /**
+     * `postAutofill` gives features an opportunity to perform an action directly
+     * following an autofill.
+     *
+     * For example, if a generated password was used, we want to fire a save event.
+     *
+     * @param {IdentityObject|CreditCardObject|CredentialsObject} data
+     * @param {DataStorageObject} formValues
+     */
+    postAutofill (data, formValues) {
+        if (AUTOGENERATED_KEY in data && 'password' in data) {
+            if (formValues.credentials?.password === data.password) {
+                const withAutoGeneratedFlag = appendGeneratedId(formValues, data.password)
+                this.storeFormData(withAutoGeneratedFlag)
+            }
+        }
+    }
+
+    /**
+     * `postSubmit` gives features a one-time-only opportunity to perform an
+     * action directly after a form submission was observed.
+     *
+     * Mostly this is about storing data from the form submission, but it can
+     * also be used like in the case of Password generation, to append additional
+     * data before it's sent to be saved.
+     *
+     * @param {DataStorageObject} values
+     * @param {import("../Form/Form").Form} form
+     */
+    postSubmit (values, form) {
+        if (!form.form) return
+
+        const checks = [
+            form.shouldPromptToStoreData,
+            form.hasValues(values),
+            this.passwordGenerator.generated
+        ]
+
+        // if *any* of the checks are truthy, proceed to offer
+        if (checks.some(Boolean)) {
```

Comment:
I think this `some` had introduced a bug in the previous iteration as well. `hasValues` must be true even if `shouldPromptToStoreData` is true, but with `some` either of them can be true. What we want is:

```js
if (hasValues === true && shouldPromptToStoreData === false) do not prompt
if (hasValues === false && shouldPromptToStoreData === true) do not prompt
```